File: Algorithm/Searchin_Algorithm/Binary_Search/BinarySearch.cpp
/*
Search: Binary Search Algorithm

Time Complexity: O(n)
Space Complexity: O(1)
*/

// #include <bits/stdc++.h>
#include <iostream>
#include <vector>
using namespace std;

bool binarySearch(vector<int> arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while(left <= right) {
        int mid = (left + right) / 2;
        if(arr[mid] == target) return true;
        if(arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return false;
}

int main() {

    vector<int> arr = {1, 2, 3, 4, 5, 6, 7 };

    cout << (binarySearch(arr, 5) ? "Found" : "Not Found") << endl;
    cout << (binarySearch(arr, 20) ? "Found" : "Not Found") << endl;

    return 0;
}

File: Data_Structure/Array/BinarySearch.java
/*
Array: Search in an Array
Search: Binary Search

Time Complexity: O(log(n))
Space Complexity: O(1)
*/

package Data_Structure.Array;

public class BinarySearch {

    public static boolean search(int[] arr, int target) {
        int left = 0;
        int right = arr.length -1;
        while (left <= right) {
            int mid = (left + right ) / 2;
            if (arr[mid] == target) {
                return true;
            }
            if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        int[] arr = {2 , 6, 19, 21, 32, 56, 86, 98};
        
        if (search(arr, 56)) {
            System.out.println("target found");
        } else {
            System.out.println("target not found");
        }
    }
}
